# GLOBAL UNDERNUTRITION STUDY - EXPLORATION  

*BY FURAWA*  

**Table of Contents**  

1. [Data collection](#data_collection)  
2. [Data discovery](#data_discovery)  
3. [Data cleaning](#data_cleaning)  
4. [Computing new variables to lead the analysis](#new_variables)  
5. [Identify major trends](#major_trends)  

In [1]:
# Import all the needed libraries for the project
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from glob import glob
import re
from pandas.api.types import CategoricalDtype

%matplotlib inline
pd.set_option('max_rows', 20)

<a id='data_collection'></a>
## 1. Data collection  
All the data has been downloaded from the [FAO](http://www.fao.org/faostat/en/#data) website.  
Let us check the files.

In [2]:
# Store the file names in the file_names variable
file_names = glob('files/*.csv')
# Check the file_names list
file_names

['files/food_balance_cereals.csv',
 'files/food_security_indicators.csv',
 'files/food_balance_vegetal.csv',
 'files/food_balance_animal.csv',
 'files/population.csv']

In [3]:
# Loop into the file_names list
for file in file_names:
    # Read each file in the file_names list and assign it to a variable retrieved from the file name
     exec(re.split('\. |\W', file)[1] + "=  pd.read_csv(file)")

In [4]:
# Check the dataframe info
food_security_indicators.head(2)

Domain Code                             Domain  Area Code         Area  \
0          FS  Suite of Food Security Indicators          2  Afghanistan   
1          FS  Suite of Food Security Indicators          2  Afghanistan   

   Element Code Element  Item Code  \
0          6132   Value     210011   
1          6132   Value     210011   

                                                Item  Year Code       Year  \
0  Number of people undernourished (million) (3-y...   20122014  2012-2014   
1  Number of people undernourished (million) (3-y...   20132015  2013-2015   

       Unit Value Flag Flag Description  Note  
0  millions   7.9    F     FAO estimate   NaN  
1  millions   8.8    F     FAO estimate   NaN

In [5]:
# Check the dataframe info
food_balance_vegetal.head(2)

Domain Code               Domain  Country Code      Country  Element Code  \
0         FBS  Food Balance Sheets             2  Afghanistan          5511   
1         FBS  Food Balance Sheets             2  Afghanistan          5611   

           Element  Item Code                Item  Year Code  Year  \
0       Production       2511  Wheat and products       2013  2013   
1  Import Quantity       2511  Wheat and products       2013  2013   

          Unit   Value Flag   Flag Description  
0  1000 tonnes  5169.0    S  Standardized data  
1  1000 tonnes  1173.0    S  Standardized data

In [6]:
# Check the dataframe info
food_balance_animal.head(2)

Domain Code               Domain  Country Code      Country  Element Code  \
0         FBS  Food Balance Sheets             2  Afghanistan          5511   
1         FBS  Food Balance Sheets             2  Afghanistan          5611   

           Element  Item Code         Item  Year Code  Year         Unit  \
0       Production       2731  Bovine Meat       2013  2013  1000 tonnes   
1  Import Quantity       2731  Bovine Meat       2013  2013  1000 tonnes   

   Value Flag   Flag Description  
0  134.0    S  Standardized data  
1    6.0    S  Standardized data

In [7]:
# Check the dataframe info
food_balance_cereals.head(2)

Domain Code               Domain  Country Code      Country  Element Code  \
0         FBS  Food Balance Sheets             2  Afghanistan          5511   
1         FBS  Food Balance Sheets             2  Afghanistan          5611   

           Element  Item Code                Item  Year Code  Year  \
0       Production       2511  Wheat and products       2013  2013   
1  Import Quantity       2511  Wheat and products       2013  2013   

          Unit   Value Flag   Flag Description  
0  1000 tonnes  5169.0    S  Standardized data  
1  1000 tonnes  1173.0    S  Standardized data

In [8]:
population.head(2)

Domain Code               Domain  Country Code      Country  Element Code  \
0         FBS  Food Balance Sheets             2  Afghanistan           511   
1         FBS  Food Balance Sheets             3      Albania           511   

                         Element  Item Code        Item  Year Code  Year  \
0  Total Population - Both sexes       2501  Population       2013  2013   
1  Total Population - Both sexes       2501  Population       2013  2013   

           Unit  Value Flag Flag Description  
0  1000 persons  30552  NaN    Official data  
1  1000 persons   3173  NaN    Official data

Except from the food_security_indicators dataframe all the other dataframes have the same 14 columns.  

<a id='data_discovery'></a>
## 2. Data Discovery

Let us check the primary key of each table and test them.  
We will create a function to find the primary key.  

In [9]:
# Function to find the potential primary keys
def check_potential_primary_key(df) -> str:
    # Loop in the column list of the specific dataframe
    for column_pk in df.keys():
        # Remove the duplicated values from each column and check if the size is the same as the df
        if len(df) != len(df[column_pk].drop_duplicates()):
            # No output if the column is not a primary key
           None
        else:
            # Print all the potential primary keys
            print("{} could be a primary key!".format(column_pk))

Now we can use the function to find the potential primary keys of each dataframe. 

In [10]:
# Check the primary key of population dataframe
check_potential_primary_key(population)

Country Code could be a primary key!
Country could be a primary key!
Value could be a primary key!


We have 3 potential primary keys for the population dataframe, but the best choice here is the Country Code variable. It won't be a good idea to have the population or country name as primary key as they will be difficult to query.

In [11]:
# Check the primary key of food balance vegetal
check_potential_primary_key(food_balance_vegetal)

We have no output, which means there are no potential primary key in this dataframe.

In [12]:
# Check the primary key of food balance livestock
check_potential_primary_key(food_balance_animal)

Same here, there are no primary keys in the food balance livestock

In [13]:
# Check the primary key of food balance cereals
check_potential_primary_key(food_balance_cereals)

In [14]:
# Check the primary key of food security indicators
check_potential_primary_key(food_security_indicators)

Even for the food balance cereals and food security indicators dataframes we have no potential primary keys

Let us create column with the total population and remove some useless columns from the population dataframe.

In [15]:
# Create the population column, we retrieve the 1000 in the Unit column and multiply it by the Value column
population["population"] = int(population.Unit.str.split(" ")[0][0]) * population.Value
# Remove some useless columns
population_df = population.drop(population.columns.difference(["Country Code", "Country", "population"]),
                                axis =1)
# Check the dataframe
population_df.head()

Country Code              Country  population
0             2          Afghanistan    30552000
1             3              Albania     3173000
2             4              Algeria    39208000
3             7               Angola    21472000
4             8  Antigua and Barbuda       90000

Now we can calculate the total numbers of human involved.

In [16]:
# Calcalute the total number of humans 
total_population = population_df.population.sum()
print("The total number of humans on the planet is : {:,}".format(total_population))

The total number of humans on the planet is : 8,413,993,000


This result cannot be correct, mostly if we talk about the 2013 world population. actually in 2019 the world population is around 7.7 billion. There must be an error, we will go deep to check the issue.  

<a id='data_cleaning'></a>
## 3. Data Cleaning  

The dataframes are downloaded and loaded but dirty. There are useless rows and columns, anomalies in the population data must be corrected, the columns names must be changed. Let's do some cleaning.  
We start putting all the food balance dataframes in one unique dataframe.  
- **Food balance**

In [17]:
# Create the origin variable in each balance food dataframe to store the food origin
food_balance_animal["origin"] = "animal"
food_balance_cereals["origin"] = "cereal"
food_balance_vegetal["origin"] = "vegetal"

In [18]:
# Append the 3 dataframes in one unique dataframe
food_balance_df = food_balance_animal.append(food_balance_vegetal)
# Check the first rows
food_balance_df.head(2)

Domain Code               Domain  Country Code      Country  Element Code  \
0         FBS  Food Balance Sheets             2  Afghanistan          5511   
1         FBS  Food Balance Sheets             2  Afghanistan          5611   

           Element  Item Code         Item  Year Code  Year         Unit  \
0       Production       2731  Bovine Meat       2013  2013  1000 tonnes   
1  Import Quantity       2731  Bovine Meat       2013  2013  1000 tonnes   

   Value Flag   Flag Description  origin  
0  134.0    S  Standardized data  animal  
1    6.0    S  Standardized data  animal

In [19]:
# Check the last rows
food_balance_df.tail(2)

Domain Code               Domain  Country Code Country  Element Code  \
104869         FBS  Food Balance Sheets           351   China           674   
104870         FBS  Food Balance Sheets           351   China           684   

                                       Element  Item Code           Item  \
104869  Protein supply quantity (g/capita/day)       2899  Miscellaneous   
104870      Fat supply quantity (g/capita/day)       2899  Miscellaneous   

        Year Code  Year          Unit  Value Flag Flag Description   origin  
104869       2013  2013  g/capita/day   0.01   Fc  Calculated data  vegetal  
104870       2013  2013  g/capita/day   0.00   Fc  Calculated data  vegetal

In [20]:
# Delete the 3 useless balance food dataframe
del food_balance_animal, food_balance_vegetal

In [21]:
# Rename the columns
food_balance_df.rename(columns = {"Country Code":"country_code", "Country":"country", "Element":"element",
                                  "Item Code":"item_code", "Item":"item", "Year":"year","Unit":"unit",
                                  "Value":"value"}, inplace = True)

In [22]:
# Transform the dataframe from long to wide with pivot_table
food_balance_wide = food_balance_df.pivot_table(
    # Put as index the Columns that we want to keep in the dataframe
    index = ["country_code", "country", "item_code", "item", "year", "unit", "origin"],
    # Select the columns that we want to transform from long to wide and the values that we sum 
    columns = ["element"], values = ["value"], aggfunc = sum)
# Renaming the columns 
food_balance_wide.columns = ["domestic_supply_quantity", "export_quantity", "fat_supply_quantity_gcapitaday",
                             "feed", "food", "food_supply_kcalcapitaday", "food_supply_quantity_kgcapitayr", 
                            "import_quantity", "losses", "other_uses", "processing", "production", 
                            "protein_supply_quantity_gcapitaday", "seed", "stock_variation"]

In [23]:
# Reset the index to have normal columns
food_balance = food_balance_wide.reset_index()
# delete the useless dataframe
del food_balance_df, food_balance_wide
# Check the first rows of the dataframe
food_balance.head()

country_code  country  item_code                 item  year  \
0             1  Armenia       2511   Wheat and products  2013   
1             1  Armenia       2511   Wheat and products  2013   
2             1  Armenia       2511   Wheat and products  2013   
3             1  Armenia       2511   Wheat and products  2013   
4             1  Armenia       2513  Barley and products  2013   

              unit   origin  domestic_supply_quantity  export_quantity  \
0      1000 tonnes  vegetal                     554.0              1.0   
1     g/capita/day  vegetal                       NaN              NaN   
2  kcal/capita/day  vegetal                       NaN              NaN   
3               kg  vegetal                       NaN              NaN   
4      1000 tonnes  vegetal                     198.0              0.0   

   fat_supply_quantity_gcapitaday  ...  food_supply_kcalcapitaday  \
0                             NaN  ...                        NaN   
1                             3.6  ...                        NaN   
2                             NaN  ...                     1024.0   
3                             NaN  ...                        NaN   
4                             NaN  ...                        NaN   

   food_supply_quantity_kgcapitayr  import_quantity  losses  other_uses  \
0                              NaN            361.0    32.0         0.0   
1                              NaN              NaN     NaN         NaN   
2                              NaN              NaN     NaN         NaN   
3                            130.6              NaN     NaN         NaN   
4                              NaN              9.0    15.0        26.0   

   processing  production  protein_supply_quantity_gcapitaday  seed  \
0        10.0       312.0                                 NaN  30.0   
1         NaN         NaN                               30.52   NaN   
2         NaN         NaN                                 NaN   NaN   
3         NaN         NaN                                 NaN   NaN   
4         7.0       189.0                                 NaN  14.0   

   stock_variation  
0           -118.0  
1              NaN  
2              NaN  
3              NaN  
4              0.0  

[5 rows x 22 columns]

Looking at the definitions of the elements in the [FAO](http://www.fao.org/faostat/en/#data/FBS)(Definitions and Standards tab), we notice that there are redundant information concerning those elements. Let's indentify the redundancy with a mathematical formula and give and example with the wheat in France.


In [24]:
# Create a dataframe with France as country and wheat as item
wheat_france = food_balance.query("country == 'France' and item == 'Wheat and products'")
# Formulas
print("Formula 1 : Domestic supply = Production + Import Quantity + Stock Variation - Export Quantity \n\
Formula 2 : Domestic supply = Food + Feed + Losses + Seed + Processing + Other Uses")

Formula 1 : Domestic supply = Production + Import Quantity + Stock Variation - Export Quantity 
Formula 2 : Domestic supply = Food + Feed + Losses + Seed + Processing + Other Uses


In [25]:
# Apply the formula in the wheat france dataframe
term_1 = (wheat_france[:1].production + wheat_france[:1].import_quantity + wheat_france[:1].stock_variation \
         - wheat_france[:1].export_quantity).values[0]

term_2 = wheat_france[:1].domestic_supply_quantity.values[0]

term_3 = (wheat_france[:1].food + wheat_france[:1].feed + wheat_france[:1].losses + wheat_france[:1].seed + \
         wheat_france[:1].processing + wheat_france[:1].other_uses).values[0]

In [26]:
# Check if all the term are equal, no output means it is correct
assert term_1 == term_2 == term_3

In [27]:
print("For the wheat in France we have : \n\
Domestic supply quantity = {} ktonnes \n\
Production + Import Quantity + Stock Variation - Export Quantity = {} ktonnes \n\
Food + Feed + Losses + Seed + Processing + Other Uses = {} ktonnes".format(term_2, term_1, term_3))

For the wheat in France we have : 
Domestic supply quantity = 20298.0 ktonnes 
Production + Import Quantity + Stock Variation - Export Quantity = 20298.0 ktonnes 
Food + Feed + Losses + Seed + Processing + Other Uses = 20298.0 ktonnes


- **Food Security Indicators**

In [28]:
food_security_indicators.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 15 columns):
Domain Code         1020 non-null object
Domain              1020 non-null object
Area Code           1020 non-null int64
Area                1020 non-null object
Element Code        1020 non-null int64
Element             1020 non-null object
Item Code           1020 non-null int64
Item                1020 non-null object
Year Code           1020 non-null int64
Year                1020 non-null object
Unit                1020 non-null object
Value               605 non-null object
Flag                1020 non-null object
Flag Description    1020 non-null object
Note                0 non-null float64
dtypes: float64(1), int64(4), object(10)
memory usage: 119.7+ KB


- The Value column has 605 non-null rows which means that there are many null values  
- There are many useless columns to remove  
- We need to change the columns name
- The Value must be float number not a string  
- The year must be a categorical variable not a string 
- There is a value <0.1, that must be changed

Let us make all those changes.

In [29]:
food_security_indicators[food_security_indicators.Value.isnull()][:5]

Domain Code                             Domain  Area Code            Area  \
15          FS  Suite of Food Security Indicators          5  American Samoa   
16          FS  Suite of Food Security Indicators          5  American Samoa   
17          FS  Suite of Food Security Indicators          5  American Samoa   
18          FS  Suite of Food Security Indicators          5  American Samoa   
19          FS  Suite of Food Security Indicators          5  American Samoa   

    Element Code Element  Item Code  \
15          6132   Value     210011   
16          6132   Value     210011   
17          6132   Value     210011   
18          6132   Value     210011   
19          6132   Value     210011   

                                                 Item  Year Code       Year  \
15  Number of people undernourished (million) (3-y...   20122014  2012-2014   
16  Number of people undernourished (million) (3-y...   20132015  2013-2015   
17  Number of people undernourished (million) (3-y...   20142016  2014-2016   
18  Number of people undernourished (million) (3-y...   20152017  2015-2017   
19  Number of people undernourished (million) (3-y...   20162018  2016-2018   

        Unit Value Flag    Flag Description  Note  
15  millions   NaN   NV  Data not available   NaN  
16  millions   NaN   NV  Data not available   NaN  
17  millions   NaN   NV  Data not available   NaN  
18  millions   NaN   NV  Data not available   NaN  
19  millions   NaN   NV  Data not available   NaN

All these rows with NaN values are useless. We will remove them. 

In [30]:
# Remove all NaN rows from the dataframe
food_security_indicators = food_security_indicators[food_security_indicators.Value.isnull() == False]

In [31]:
# Check that the Value column has no more NaN values, no output means it is correct
assert food_security_indicators.Value.isnull().all() == False

Now we can change the columns name and keep the columns that we need for the Analysis.

In [74]:
# Select the columns that we need
indicators_df = food_security_indicators.loc[:, ["Area Code", "Area", "Year", "Value"]]
# Change the columns name
indicators_df.columns = ["country_code", "country", "year", "value"]
# Turn the year variable into a categorical type
category_type = CategoricalDtype(categories = ["2012-2014", "2013-2015", "2014-2016", "2015-2017", "2016-2018"]
                                               ,ordered = True)
indicators_df.year = indicators_df.year.astype(category_type)
indicators_df.year = indicators_df.year.cat.rename_categories([2013, 2014, 2015, 2016, 2017])

In [75]:
# Assert that the data type is correct
assert indicators_df.year.dtype == "category"

In [78]:
# Replace the <0.1 value with 0.09 
indicators_df.value = indicators_df.value.apply(lambda x : x.replace("<0.1", "0.09"))

In [77]:
# Check if there is no <0.1 value anymore
assert indicators_df.loc[indicators_df.value == "<0.1"].value.any() == False

In [79]:
# Change the data type of the value columns from string to float
indicators_df.value = indicators_df.value.astype("float64")

In [80]:
# Assert that the changes occurs
assert indicators_df.value.dtype == "float64"

In [82]:
indicators_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 605 entries, 0 to 1019
Data columns (total 4 columns):
country_code    605 non-null int64
country         605 non-null object
year            605 non-null category
value           605 non-null float64
dtypes: category(1), float64(1), int64(1), object(1)
memory usage: 19.7+ KB


In [83]:
indicators_df.sample(3)

country_code country  year  value
28              7  Angola  2016    7.4
374            81   Ghana  2017    1.6
873           276   Sudan  2016    7.9

- **Population**  

There is one flag value. Let see what it is about

In [96]:
population[population.Flag.isnull() == False]

Domain Code               Domain  Country Code Country  Element Code  \
174         FBS  Food Balance Sheets           351   China           511   

                           Element  Item Code        Item  Year Code  Year  \
174  Total Population - Both sexes       2501  Population       2013  2013   

             Unit    Value Flag  \
174  1000 persons  1416667    A   

                                      Flag Description  population  
174  Aggregate, may include official, semi-official...  1416667000

We see here that The "China" entry as country is an aggregate value. Let's check if there are another entry with China.

In [101]:
# Find all the entries with china
population_df[population_df.Country.str.contains("China")]

Country Code                    Country  population
32             96       China, Hong Kong SAR     7204000
33            128           China, Macao SAR      566000
34             41            China, mainland  1385567000
35            214  China, Taiwan Province of    23330000
174           351                      China  1416667000

In [103]:
# Sum the four first values
population_df.population[32:36].sum()

1416667000

We can see that China, Hong Kong SAR, China, Macao SAR, China, mainland and China, Taiwan Province of are actually four parts of the China country. We can confirm this because the total population of the four parts are equal to the population of the China country.
We will remove the China population value from the dataframe during the calculation of the total number of humans on the planet.

In [107]:
# Remove China from the dataframe
population_df = population_df[population_df.Country != "China"]

In [109]:
# Compute the wold population
world_population = population_df.population.sum()
print("The total number of humans on the planet in 2013 is : {:,}".format(world_population))

The total number of humans on the planet in 2013 is : 6,997,326,000
